src/Database/LSMTree/Internal.hs

```diff
@@ -334,20 +415,31 @@ new ::
   -> TableConfig
   -> m (TableHandle m h)
 new sesh conf = do
+    -- Keep the session open until we've updated the session's set of tracked
+    -- tables. If 'closeSession' is called by another thread while this code
+    -- block is being executed, that thread will block until it reads the
+    -- /updated/ set of tracked tables.
     withOpenSession sesh $ \seshEnv -> do
       writeBufVar <- newMVar (WB.empty)
       levelsVar <- newMVar V.empty
       cacheVar <- newMVar (mkLevelsCache V.empty)
+      tableId <- incrUniqCounter (sessionUniqCounter seshEnv)
+      -- Action to untrack the current table
+      let forget = modifyMVar_ (sessionOpenTables seshEnv) $ pure . Map.delete tableId
```

Comment:
This action will access the MVar without considering the session RW lock, but it should be fine. If session and table get closed concurrently, this might mean that the table gets closed twice, but it's idempotent. :+1: